lib/edgehog_device_forwarder/connections/token_data.ex

```diff
@@ -0,0 +1,28 @@
+# Copyright 2023 SECO Mind Srl
+# SPDX-License-Identifier: Apache-2.0
+
+defmodule EdgehogDeviceForwarder.Connections.TokenData do
+  defstruct realm: nil, device_id: nil, socket: nil
+
+  @type t :: %__MODULE__{}
+  @type device_id :: String.t()
+  @type realm :: String.t()
+  @type socket :: pid()
+
+  @spec new(socket) :: t
+  def new(socket), do: %__MODULE__{socket: socket}
+
+  @spec put_device_info(nil | t, %{:device_id => device_id, :realm => realm}) ::
+          {:ok, t} | {:error, :token_not_found}
+  def put_device_info(nil, _device_info), do: {:error, :token_not_found}
+
+  def put_device_info(data, device_info) do
```

Comment:
I'd expect `put_device_info` to guard explicitly on `%TokenData{}` and just crash if it gets passed `nil`. That's how all functions acting on a struct usually work (see e.g. all the functions in the `Ecto.Changeset` module, or `Plug.Conn`).

This also saves you from returning an error tuple since the result is guaranteed to succeed.

Reply:
I can understand not wanting the function to return `:ok | :error` tuple, but I personally think that adding explicit type signature in our internal modules is unnecessary boilerplate.

as these functions are called inside our "safe" modules (with type/error checking done at the api boundaries), typespecs (and adding dialyzer in ci) should be enough to guarantee the usage is consistent with what we expect

Reply:
Looking at the code, I probably did the `:ok | :error` thing because it works really well for `ConCache.update/3` [here](https://github.com/edgehog-device-manager/edgehog_device_forwarder/pull/6/commits/8633b47be280bfced141230c7a849b875edcb0b8#diff-2ffe2dd5cedfa2908db334b50c97fb654ae444ea1178f4a0f9abd3b4ba0d0a3aR24), although not the best solution for the project structure